### PR TITLE
Issue #1019: Add support for tags at vpc module

### LINKED ISSUE #1019: Add support for tags at vpc module
Include addition and update of tags

### PR DESCRIPTION
**Pull request by yrobla**
Include adding and updated tags

Signed-off-by: Yolanda Robla <info@ysoft.biz>
Resolves #1019

### CODE REVIEW

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 1]
```diff
@@ -543,7 +567,11 @@ export const AwsVpcModule: Module2 = new Module2({
               await AwsVpcModule.mappers.vpc.db.update(cloudRecord, ctx);
               out.push(cloudRecord);
             } else {
-              await AwsVpcModule.mappers.vpc.cloud.create(e, ctx);
+              const res = await AwsVpcModule.mappers.vpc.cloud.create(e, ctx);
+              if (!AwsVpcModule.utils.eqTags(e.tags, cloudRecord.tags) && e.vpcId && e.tags) {
+                await updateTags(client.ec2client, res.vpcId, e.tags);
+              }
+
```
Comment (aguillenv) [inline src/modules/0.0.14/aws_vpc/index.ts:574]: oh wait you should not be doing this here, this is the replace path. Also, please add the `updateOrReplace` mapper function to identify the cases (we forgot in the previous PRs). Then, the tags' update and the current restore should go through the `update` path.

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 2]
```diff
@@ -359,6 +359,12 @@ export const AwsVpcModule: Module2 = new Module2({
       out.cidrBlock = vpc.CidrBlock;
       out.state = vpc.State as VpcState;
       out.isDefault = vpc.IsDefault ?? false;
+      const tags: { [key: string]: string } = {};
+      (vpc.Tags || []).filter(t => !!t.Key && !!t.Value).forEach(t => {
```
Comment (dfellis) [inline src/modules/0.0.14/aws_vpc/index.ts:363]: Nit. This filter would technically filter out a key or value that is an empty string. Probably fine for the key, but I could see someone legitimately using just the presence of a key in a HashSet-style and not care about the value, and this would drop it.

Perhaps this would be better:

```suggestion
      (vpc.Tags || []).filter(t => t.hasOwnProperty('Key') && t.hasOwnProperty('Value')).forEach(t => {
```

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 3]
```diff
@@ -536,15 +564,24 @@ export const AwsVpcModule: Module2 = new Module2({
           const client = await ctx.getAwsClient() as AWS;
           const out = [];
           for (const e of es) {
-            const cloudRecord = ctx?.memo?.cloud?.Vpc?.[e.vpcId ?? ''];
+            const cloudRecord = ctx?.memo?.cloud?.Vpc?.[e.vpcId ?? ''];
```
Comment (aguillenv) [inline src/modules/0.0.14/aws_vpc/index.ts:567]:
```suggestion
            const cloudRecord = ctx?.memo?.cloud?.Vpc?.[e.vpcId ?? ''];
```

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 4]
```diff
@@ -495,19 +501,41 @@ export const AwsVpcModule: Module2 = new Module2({
...
       equals: (a: Vpc, b: Vpc) => {
-        const result = Object.is(a.cidrBlock, b.cidrBlock) && Object.is(a.state, b.state) && Object.is(a.isDefault, b.isDefault);
+        const result = Object.is(a.cidrBlock, b.cidrBlock) && Object.is(a.state, b.state) && Object.is(a.isDefault, b.isDefault) &&
+          (AwsVpcModule.utils.eqTags(a.tags, b.tags));
         return result;
       },
       source: 'db',
       cloud: new Crud2({
+        updateOrReplace: (a: Vpc, b: Vpc) => {
+          if (Object.is(a.cidrBlock, b.cidrBlock)) return "replace";
```
Comment (aguillenv) [inline src/modules/0.0.14/aws_vpc/index.ts:511]: Hmm I think here it is the opposite, right? 
```suggestion
          if (!Object.is(a.cidrBlock, b.cidrBlock)) return "replace";
```

Comment (aguillenv) [reply src/modules/0.0.14/aws_vpc/index.ts:511]: Please also include the `isDefault` property for the `replace` path

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 5]
```diff
@@ -419,8 +425,13 @@ export const AwsVpcModule: Module2 = new Module2({
       if (serviceName.includes('s3')) return EndpointGatewayService.S3;
       if (serviceName.includes('dynamodb')) return EndpointGatewayService.DYNAMODB;
     },
-    eqTags: (a: { [key: string]: string }, b: { [key: string]: string }) => Object.is(Object.keys(a ?? {})?.length, Object.keys(b ?? {})?.length) &&
-      Object.keys(a ?? {})?.every(ak => (a ?? {})[ak] === (b ?? {})[ak]),
+    eqTags: (a: { [key: string]: string }, b: { [key: string]: string }) => {
+      // remove possible `updated` key
+      delete a.updated;
+      delete b.updated;
```
Comment (aguillenv) [inline src/modules/0.0.14/aws_vpc/index.ts:431]: oh why are you doing this?

Comment (dfellis) [reply src/modules/0.0.14/aws_vpc/index.ts:431]: Agree here. This shouldn't be necessary.

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 6]
```diff
@@ -495,20 +506,45 @@ export const AwsVpcModule: Module2 = new Module2({
...
         create: async (es: Vpc[], ctx: Context) => {
           // TODO: Add support for creating default VPCs (only one is allowed, also add constraint
           // that a single VPC is set as default)
           const client = await ctx.getAwsClient() as AWS;
           const out = [];
           for (const e of es) {
+            let tgs: Tag[] = [];
+            if (e.tags !== undefined && e.tags !== null) {
+              const tags: {[key: string]: string} = e.tags;
+              tags.owner = 'iasql-engine';
```
Comment (dfellis) [inline src/modules/0.0.14/aws_vpc/index.ts:528]: Why is this here?

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 7]
```diff
@@ -495,20 +506,45 @@ export const AwsVpcModule: Module2 = new Module2({
...
+              tgs = Object.keys(tags).map(k => {
+                return {
+                  Key: k, Value: tags[k]
+                }
+              });
```
Comment (dfellis) [inline src/modules/0.0.14/aws_vpc/index.ts:533]: Could also be

```suggestion
              tgs = Object.entries(tags).map(([Key, Value]) => ({ Key, Value, }));
```

File: src/modules/0.0.14/aws_vpc/index.ts [hunk 8]
```diff
@@ -543,8 +579,17 @@ export const AwsVpcModule: Module2 = new Module2({
...
             } else {
-              await AwsVpcModule.mappers.vpc.cloud.create(e, ctx);
-              out.push(e);
+              const isUpdate = Object.is(AwsVpcModule.mappers.vpc.cloud.updateOrReplace(cloudRecord, e), 'update');
+              if (!isUpdate) {
+                // if CIDR is different we do the create
+                const newVpc = await AwsVpcModule.mappers.vpc.cloud.create(e, ctx);
+                out.push(newVpc);
+              } else {
+                if (!AwsVpcModule.utils.eqTags(e.tags, cloudRecord.tags) && e.vpcId && e.tags) {
```
Comment (dfellis) [inline src/modules/0.0.14/aws_vpc/index.ts:588]: I would remove this last `&& e.tags` it should still update the tags if the user has deleted all of them.

```suggestion
                if (!AwsVpcModule.utils.eqTags(e.tags, cloudRecord.tags) && e.vpcId) {
```

File: test/modules/aws-vpc-integration.ts [hunk 9]
```diff
@@ -101,6 +105,25 @@ describe('VPC Integration Testing', () => {
     AND state='pending';
   `, (res: any) => expect(res.length).toBe(0)));
 
+  it('tries to update vpc tags', query(`
+  UPDATE vpc SET tags = '{"name": "${prefix}-2", "updated": true}' WHERE cidr_block='192.${randIPBlock}.0.0/16';
```
Comment (dfellis) [inline test/modules/aws-vpc-integration.ts:109]:
```suggestion
  UPDATE vpc SET tags = '{"name": "${prefix}-2"}' WHERE cidr_block='192.${randIPBlock}.0.0/16';
```

**Review by dfellis — APPROVED** [review]

**Review by aguillenv — APPROVED** [review]
